bittide-instances/src/Project/Handle.hs

```diff
@@ -0,0 +1,70 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Project.Handle where
+
+import Prelude
+
+import Data.List.Extra (trim)
+import System.IO
+
+import Test.Tasty.HUnit
+
+data Error = Ok | Error String
+data Filter = Continue | Stop Error
+
+{- | Utility function that reads lines from a handle, and applies a filter to
+each line. If the filter returns 'Continue', the function will continue
+reading lines. If the filter returns @Stop Ok@, the function will return
+successfully. If the filter returns @Stop (Error msg)@, the function will
+fail with the given message.
+-}
+expectLine :: (HasCallStack) => Handle -> (String -> Filter) -> Assertion
+expectLine h f = do
+  line <- trim <$> hGetLine h
+  let cont = expectLine h f
+  if null line
+    then cont
+    else case f line of
+      Continue -> cont
+      Stop Ok -> pure ()
+      Stop (Error msg) -> assertFailure msg
+
+{- | Utility function that reads lines from a handle, and waits for a specific
+line to appear. Though this function does not fail in the traditional sense,
+it will get stuck if the expected line does not appear. Only use in combination
+with sensible time outs (also see 'main').
+-}
+waitForLine :: Handle -> String -> IO ()
+waitForLine h expected =
+  expectLine h $ \s ->
+    if s == expected
+      then Stop Ok
+      else Continue
+
+readRemainingChars :: Handle -> IO String
+readRemainingChars h = do
+  hSetBuffering h NoBuffering
+  rdy <- hReady h
+  if rdy
+    then do
+      c <- hGetChar h
+      (c :) <$> readRemainingChars h
+    else (pure "")
+
+readRemainingLines :: Handle -> IO [String]
+readRemainingLines h = do
+  hSetBuffering h LineBuffering
+  rdy <- hReady h
+  if rdy
+    then do
+      c <- hGetLine h
+      (c :) <$> readRemainingLines h
+    else pure []
```

Comment:
@leonschoorl What do you think about those implicit `hSetBuffering` calls? They seem really iffy to me.

Reply:
In my mind those buffering modes are kind of set-and-forget, you set it once and then never change then again.
But if you have functions that need specific modes, maybe you should read out the current mode at the start of the function, change it, do your thing, and restore the previous mode at the end.
That feels like a clean way of handling it.

But then again, if the only functions in the codebase touching those handles are these two, then if feels a little unnecessary to be restoring the modes all the time.